transport/x/grpchttp/grpchttp.go

```diff
@@ -0,0 +1,227 @@
+func HTTPMux(pattern string, serveMux *http.ServeMux) InboundOption {
+	return func(inboundOptions *inboundOptions) {
+		inboundOptions.httpInboundOptions = append(inboundOptions.httpInboundOptions, yarpchttp.Mux(pattern, serveMux))
+	}
+}
+
+// NewInbound returns a new Inbound.
+func NewInbound(listener net.Listener, options ...InboundOption) *Inbound {
+	return newInbound(listener, options...)
+}
+
+// Inbound is a transport.Inbound.
+type Inbound struct {
+	once     *lifecycle.Once
+	lock     sync.Mutex
+	listener net.Listener
+	options  *inboundOptions
+	router   transport.Router
+
+	grpcTransport *yarpcgrpc.Transport
+	grpcInbound   *yarpcgrpc.Inbound
+	httpTransport *yarpchttp.Transport
+	httpInbound   *yarpchttp.Inbound
+}
+
+func newInbound(listener net.Listener, options ...InboundOption) *Inbound {
+	return &Inbound{
+		once:     lifecycle.NewOnce(),
+		listener: listener,
+		options:  newInboundOptions(options),
+	}
+}
+
+// Start implements transport.Lifecycle#Start.
+func (i *Inbound) Start() error {
+	return i.once.Start(i.start)
+}
+
+// Stop implements transport.Lifecycle#Stop.
+func (i *Inbound) Stop() error {
+	return i.once.Stop(i.stop)
+}
+
+// IsRunning implements transport.Lifecycle#IsRunning.
+func (i *Inbound) IsRunning() bool {
+	return i.once.IsRunning()
+}
+
+// SetRouter implements transport.Inbound#SetRouter.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.router = router
+}
+
+// Transports implements transport.Inbound#Transports.
+func (i *Inbound) Transports() []transport.Transport {
+	return []transport.Transport{}
+}
+
+func (i *Inbound) start() error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	if i.router == nil {
+		return errRouterNotSet
+	}
+
+	mux := cmux.New(i.listener)
+	// TODO: Using MatchWithWriters/SendSettings is a major performance hit
+	// Per the cmux documentation, you have to do this for grpc-java
+	// Since we are only using grpc-go, this might be done elsewhere
+	// Need to look into this further
+	grpcListener := mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldPrefixSendSettings("content-type", "application/grpc"))
+	httpListener := mux.Match(cmux.Any())
+	mux.HandleError(func(err error) bool {
+		if err != cmux.ErrListenerClosed {
+			zap.L().Error(err.Error())
+		}
+		return false
+	})
+
+	grpcTransport := yarpcgrpc.NewTransport(i.options.grpcTransportOptions...)
+	grpcInbound := grpcTransport.NewInbound(grpcListener, i.options.grpcInboundOptions...)
+	grpcInbound.SetRouter(i.router)
+
+	httpTransport := yarpchttp.NewTransport(i.options.httpTransportOptions...)
+	httpInbound := httpTransport.NewInboundForListener(httpListener, i.options.httpInboundOptions...)
+	httpInbound.SetRouter(i.router)
+
+	if err := grpcTransport.Start(); err != nil {
+		_ = i.listener.Close()
+		i.listener = nil
+		return err
+	}
+	if err := grpcInbound.Start(); err != nil {
+		_ = grpcTransport.Stop()
+		_ = i.listener.Close()
```

Comment:
once Transport is added, that should be responsible for stopping the grpc and http transports.